data/api.go

```diff
@@ -0,0 +1,368 @@
+		}
+
+		if queryParams.Latest {
+			// If we're using the `latest` parameter, then we ran an `$aggregate` query to get only the latest data.
+			// Since we use Mongo 3.2, we can't use the $replaceRoot function, so we need to manually extract the
+			// latest subdocument here. When we move to MongoDB 3.4+ and can use $replaceRoot, we can get rid of this
+			// conditional block. We'd also need to fix the corresponding code in `store.go`
+			results = results["latest_doc"].(map[string]interface{})
+		}
+		if len(results) > 0 {
+			if results["type"].(string) == "pumpSettings" && parametersHistory != nil {
+				payload := results["payload"].(map[string]interface{})
+				payload["history"] = parametersHistory["history"]
+				results["payload"] = payload
+			}
+			if bytes, err := json.Marshal(results); err != nil {
+				log.Printf("%s request %s user %s Marshal returned error: %s", DATA_API_PREFIX, requestID, queryParams.UserID, err)
+			} else {
+				if writeCount > 0 {
+					res.Write([]byte(","))
+				}
+				res.Write([]byte("\n"))
+				res.Write(bytes)
+				writeCount++
+			}
+		}
+	}
+
+	if writeCount > 0 {
+		res.Write([]byte("\n"))
+	}
+	res.Write([]byte("]"))
+
+	if queryDuration := time.Now().Sub(queryStart).Seconds(); queryDuration > slowQueryDuration {
+		// XXX use metrics
+		//log.Printf("%s request %s user %s GetDeviceData took %.3fs", DATA_API_PREFIX, requestID, userID, queryDuration)
+	}
+	log.Printf("%s request %s user %s took %.3fs returned %d records", DATA_API_PREFIX, requestID, queryParams.UserID, time.Now().Sub(start).Seconds(), writeCount)
+}
+
+//log error detail and write as application/json
+func jsonError(res http.ResponseWriter, err detailedError, startedAt time.Time) {
+	logError(&err, startedAt)
+	jsonErr, _ := json.Marshal(err)
+
+	res.Header().Add("content-type", "application/json")
+	res.WriteHeader(err.Status)
+	res.Write(jsonErr)
+}
+
+func logError(err *detailedError, startedAt time.Time) {
+	err.Id = uuid.New().String()
+	log.Println(DATA_API_PREFIX, fmt.Sprintf("[%s][%s] failed after [%.3f]secs with error [%s][%s] ", err.Id, err.Code, time.Now().Sub(startedAt).Seconds(), err.Message, err.InternalMessage))
+}
+
+//set the internal message that we will use for logging
+func (d detailedError) setInternalMessage(internal error) detailedError {
+	d.InternalMessage = internal.Error()
+	return d
+}
+
+func (a *Api) userCanViewData(authenticatedUserID string, targetUserID string) bool {
+	if authenticatedUserID == targetUserID {
+		return true
+	}
+
+	perms, err := a.perms.UserInGroup(authenticatedUserID, targetUserID)
+	if err != nil {
+		log.Println(DATA_API_PREFIX, "Error looking up user in group", err)
+		return false
+	}
+
+	log.Println(perms)
+	return !(perms["root"] == nil && perms["view"] == nil)
+}
+
+func inArray(needle string, arr []string) bool {
+	for _, n := range arr {
+		if needle == n {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
This function is not used (moved now in store package which is not great... maybe we should have a dedicated package for utility functions. I may make this change when implementing new routes)

Reply:
removed